### PR TITLE
SNOW-294154 S3 PUT Support Additional Dependencies

### PR DESCRIPTION
**Pull request by sfc-gh-jbahk**
### Description
Support PUT command (stream and file) using S3 SDK, uploading to S3 stage and in turn copying to table.

### Checklist
- [x] Code compiles correctly
- [x] Run ``make fmt`` to fix inconsistent formats
- [x] Run ``make lint`` to get lint errors and fix all of them
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: storage_client.go [hunk 1]
```diff
@@ -13,9 +23,139 @@ type storageUtil interface {
...
+		}
+		if meta.resStatus == uploaded || meta.resStatus == renewToken || meta.resStatus == renewPresignedURL {
+			return nil
+		} else if meta.resStatus == needRetry {
+			lastErr = meta.lastError
+			if !meta.noSleepingTime {
+				sleepingTime := intMin(int(math.Exp2(float64(retry))), 16)
+				time.Sleep(time.Second * time.Duration(sleepingTime))
+			}
+		} else if meta.resStatus == needRetryWithLowerConcurrency {
```
Comment (sfc-gh-cshi) [inline storage_client.go:105]: where comes this error?

File: file_transfer_agent.go [hunk 2]
```diff
@@ -704,3 +767,49 @@ type snowflakeProgressPercentage struct {
 	done            bool
 	startTime       time.Time
 }
+
+func (spp *snowflakeProgressPercentage) call(bytesAmount int) {
```
Comment (sfc-gh-cshi) [inline file_transfer_agent.go:771]: Seems we don't have tests or use the progress code? In python it is used - thus tested in snowsql but we might need tests in golang implementation for ourselves.

**Review by sfc-gh-cshi — CHANGES_REQUESTED** [review]

File: file_transfer_agent.go (hunk 2 above)
Comment (sfc-gh-jbahk) [reply file_transfer_agent.go:911]: I'll be adding more tests!

**Comment by sfc-gh-hchaturvedi** [comment]
Seems like we've changed the original PR to now include only dependencies being added? Is that correct? Apart from the whitespace removal and dependency addition I don't see any other changes - just confirming if that's the goal of this PR.

**Comment by sfc-gh-jbahk** [comment]
@sfc-gh-hchaturvedi yes, that is the case! The feature implementation of the PR is here: https://github.com/snowflakedb/gosnowflake/pull/409.

**Review by sfc-gh-hchaturvedi — COMMENTED** [review]
Is it possible to move all testing related files under a single directory, including the `data` dir.

**Comment by sfc-gh-jbahk** [comment]
@sfc-gh-hchaturvedi I can look into the restructuring in a future JIRA/PR. A quick look into it wasn't as informative.

**Comment by sfc-gh-hchaturvedi** [comment]
alright in that case please rename the `data` dir to `test data` or something on similar lines.

**Comment by sfc-gh-jbahk** [comment]
@sfc-gh-hchaturvedi done!

**Review by sfc-gh-hchaturvedi — APPROVED** [review]
LGTM.

**Pull request merged**